### PR TITLE
[ci] add metacoq

### PR DESCRIPTION
**Pull request by mattam82**
If you thought we didn't exercise the kernel's interfaces and extraction enough, here's a way to do it. This adds metacoq to the CI. The build is short (~6min) but a bit intricate (it separately builds various packed ocaml libraries depending on each other) and I only tested on Linux Travis instances so it might fail on other platforms.

### CODE REVIEW

File: dev/ci/ci-metacoq.sh
Comment (SkySkimmer) [inline dev/ci/ci-metacoq.sh]: Why not make everything in one go?

Comment (mattam82) [reply dev/ci/ci-metacoq.sh]: This is a multi-make project, I didn't try but suspect it would not respect the order

**Review by ejgallego — COMMENTED** [review]
Things look OK even if a bit heavy, I assume that you folks upstream have the proper CI to keep things glued properly. A couple of comments:

- I'd much prefer if you would fix all the OCaml warnings in the build, this is still under discussion but the plan is to require plugins to adhere to the same compiler flags than the main tree for being in the CI.
- What is the plan for having metaCoq depend on released versions of the embedded libraries so they could be built from upstream?

[Maybe @Zimmi48 this is an example where having multiple repos doesn't work so well for them as to maintain coherence]

**Comment by mattam82** [comment]
Most ocaml warnings you see are generated by the compilation of extracted programs (unused opens), I can fix the rest. 

What do you mean by embedded libraries? The whole project depends on nothing else than Coq and Ocaml. We plan to release separate packages for separate functionalities but indeed, this will probably stay monorepo to maintain coherence, those are tightly knit developments.

**Comment by ejgallego** [comment]
> Most ocaml warnings you see are generated by the compilation of extracted programs (unused opens), I can fix the rest.

Oh, the extracted warnings shouldn't matter indeed.

> What do you mean by embedded libraries?

I mean the "various packed ocaml libraries", somehow I thought you were grabbing them from different repos sorry.

**Comment by Zimmi48** [comment]
>[Maybe @Zimmi48 this is an example where having multiple repos doesn't work so well for them as to maintain coherence]

I think the confusion has been cleared out but this is a case of a development using a monorepo, and since it is the same team maintaining the various packages, it is clearly a net win.

**Comment by ejgallego** [comment]
For info, I am holding this for a bit as there are several large-scope PRs [Sprop, proof_global, etc...] that require coordination with this one as they would mess the CI status.

Given that the PRs we are talking about are ready and have been opened for quite a long time they do take priority over this one in my view, but of course we can coordinate in a different way.

**Comment by ejgallego** [comment]
Note the rationale for the order works as follow: SProp, Proof_global, etc... are PR that should be merged way before the beta, so if this PR would delay them like 10 days that could be fatal.

On the other hand, this PR can be merged even in the beta phase so no such problem.

In fact, it would make sense to recommend new CI entries to be submitted in the beta period, but with 2 releases a year makes the window a bit narrow IMHO.

**Comment by Zimmi48** [comment]
> In fact, it would make sense to recommend new CI entries to be submitted in the beta period, but with 2 releases a year makes the window a bit narrow IMHO.

I'd prefer that there be no criterion restricting when people can submit PRs. It is then up to the CI maintainers to decide when is the right time to merge.

**Comment by ejgallego** [comment]
A recommendation is not restriction, but a hint, right? I guess indeed we may say nothing, I guess it will depend on each case.

**Comment by ejgallego** [comment]
Just to be clear _I propose_ to delay a bit this PR as not to interfere with the ongoing 8.10 PRs, but it is just an idea, we can merge now if you folks think that's going to work better.

**Comment by mattam82** [comment]
I'm fine with delaying. Anyway, we will continue developing MetaCoq on 8.9 branch until mid april I think. Indeed SProp should have priority. Proof_global will be far less work I anticipate :)

**Comment by ejgallego** [comment]
Almost all the important stuff is in, only waiting on #9129

**Comment by ejgallego** [comment]
No more stuff blocking this PR.

**Comment by Zimmi48** [comment]
But a rebase is needed @mattam82 @ejgallego.

**Comment by ejgallego** [comment]
> But a rebase is needed @mattam82 @ejgallego.

This is on the hands of @mattam82 and the rest of MetaCoq developers as they first need to make it work with master; I'll let them set the milestone.

**Comment by mattam82** [comment]
We'll wait until after the branching, right now we're concentrating on the 8.8/8.9 versions anyway

**Comment by ejgallego** [comment]
Cleaning up my "assigned" list, please ping when activity resumes.

File: dev/ci/ci-metacoq.sh
Comment (SkySkimmer) [inline dev/ci/ci-metacoq.sh]: It seems the `ci-local` target also does configure

Comment (mattam82) [reply dev/ci/ci-metacoq.sh]: It shouldn't, I'll fix that upstream.

**Comment by mattam82** [comment]
@SkySkimmer any idea why validate:quick failed?

**Comment by mattam82** [comment]
So this is ready I believe.

**Comment by mattam82** [comment]
@coq/ci-maintainers ?

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: Had to add this

**Comment by SkySkimmer** [comment]
The log seems a bit long for gitlab, can you make it quieter somehow?

**Comment by mattam82** [comment]
Done, it should be reasonable now

**Comment by mattam82** [comment]
I've relaunched the ci job to test that.

**Comment by mattam82** [comment]
Funny enough, it seems that rebase has broken a proof in metacoq...

**Comment by mattam82** [comment]
@SkySkimmer the pipeline actually succeeded https://gitlab.com/coq/coq/pipelines/128238914 and https://gitlab.com/coq/coq/-/jobs/479938862 I just canceled other unrelated jobs

**Comment by SkySkimmer** [comment]
Still pretty long imo, do you really need to print all those terms?

**Comment by mattam82** [comment]
I can fix the printings from the test-suite. Anyway, Coq broke it again... I'll ping when this is ready again.

**Comment by mattam82** [comment]
@SkySkimmer This should be ok now, let's wait for CI one last time

**Review by SkySkimmer — APPROVED** [review]

**Pull request merged**
